Run MLProjects on docker containers

marcusrehm opened the pull request:
This PR adds the feature to run experiments inside docker containers. Both docker image and path inside the container to put the model have to be specified in MLProject file.

It mounts local directories inside a container and runs the experiment. The project uri can be a local directory or a git url.

The attributes in MLProject can be found in projects.rst documentation.

Commit:
add ability to run training in docker containers

smurching commented:
Hi @marcusrehm, sorry for the delay in getting to this, thanks for the PR! Out of curiosity, what use cases were you trying to address via this PR? Running projects in docker containers is definitely something we want to support in MLflow, but just wanted to get a feel for the use cases / requirements you had in mind.

I also wanted to discuss the high-level design a bit before diving into the code:

* I noticed the PR adds execution in a docker container as a new mode, in addition to adding a dockerfile as an attribute of a project. Another possibility is to allow a docker image to be the execution environment for a project, and then allow different execution modes to decide how to run the project in an appropriate environment (i.e. don't have a designated "docker" execution mode). Do you have thoughts on the two options?
* To get code into the container, we could try mounting the host machine's project directory into the container (as opposed to building a new image). Is there a reason you went with building a new image?
* It looks like this PR allows projects to contain both a docker environment & conda environment. It might be simpler to allow a project to have a single environment (either docker or conda) than attempt to manage both.

Thanks again for the contribution & let me know what you think!

marcusrehm commented:
Hi @smurching ! Thank you for the feedback!

The use cases I have in mind for this PR were basically two. Run projects locally inside a docker container and so doesn't rely on conda environments and the second one is be able to deploy it on kubernetes for scale out and run it with more resources. I think both are described in [issue 302](https://github.com/mlflow/mlflow/issues/302).

Actually the idea is to have a [docker.yaml](https://github.com/mlflow/mlflow/blob/5e6124221026337234d993706c94693f33030a03/examples/docker/docker.yaml) the same way we have a conda.yaml. In this file we would have a structure like this:

``` yaml
    run:
        dockerfile:
            image_name: mlflow-run-image-file
            path: Dockerfile
        deployment_path: /opt/model
```

For running locally, it uses only the attributes `image_name` and `deployment_path` and works the same way you described, mounting the local path inside the container in the specified `deployment_path`. This container is based on `image_name` image.

About your comments regarding high-level design:
>I noticed the PR adds execution in a docker container as a new mode, in addition to adding a dockerfile as an attribute of a project. Another possibility is to allow a docker image to be the execution environment for a project, and then allow different execution modes to decide how to run the project in an appropriate environment (i.e. don't have a designated "docker" execution mode). Do you have thoughts on the two options?

The Dockerfile is just for the use case where we build the image for use in kubernetes, for example. But it could be taken out in favor of using the attribute `image_name`.

Regarding the modes, I think the first case you described is the one which is addressed as described above when running in local mode. But now that you explained the other approach I think it would be better to let the execution modes handle how they threat the docker environment. For local mode it would be as implemented, for databricks it would do nothing, I suppose. I think we can go this way. What do you think?

>To get code into the container, we could try mounting the host machine's project directory into the container (as opposed to building a new image). Is there a reason you went with building a new image?

This is just for the second use case, when we would like to deploy it on kubernetes or have a fully containerized project. For running locally it is mounting a local path as described above.

>It looks like this PR allows projects to contain both a docker environment & conda environment. It might be simpler to allow a project to have a single environment (either docker or conda) than attempt to manage both.

Yes, I pretty much agree with you. I just let it because I would like to hear you from guys about some use cases where it could be needed.

marcusrehm commented:
Hi @smurching , I did some changes and to make this PR more clear I focused only in the first (main) use case, which is the ability to run projects inside docker containers locally. I removed the feature to build image.

I refactored the code to put docker execution as a environment (like conda) and not as a mode (like databricks or local). Now if one wants to run the project inside a docker container she just needs to specify docker.yaml file and run mlflow the same way as before.

If MLProject has both conda and docker environments specified, docker takes precedence over conda.

smurching reviewed (commented):
Thanks for the updates & sorry again for the delay @marcusrehm!

Based on our earlier discussion I ended up writing a doc detailing the design options for docker project execution, if you have any thoughts I'd love to hear them: https://docs.google.com/document/d/1je4dqR4gOJlEHtUnPvtf3fVydDpwWOBN-IU_euopHeA/edit?usp=sharing

I'll discuss the major design choices with some committers tomorrow & leave more substantial feedback ASAP, for now here are some initial comments - thanks again for working on this, it'll be a great feature to have :)

marcusrehm commented on examples/docker/.gitignore:
```diff
@@ -0,0 +1,104 @@
+# Byte-compiled / optimized / DLL files
```
Actually I was versioning this example in another repo before bring it to mlflow project. I will exclude that. No worries!

marcusrehm commented on docs/source/projects.rst:
```diff
@@ -215,6 +219,19 @@ for your run. Then, run your project via
+
+.. code:: yaml
+
+    run:
```
For the `run` field I was thinking about the future, when we could have configuration for `serve` environment too. Imagine that I can train a model using Spark and so I will need it configured inside the image used for training, but for serving using MLeap for example, I wouldn't need the same image. I could use a lighter one. What do you think?

The point about having docker configuration inline MLProject. I think it's better to have a docker.yaml the same way as the conda.yaml. Right now, besides the image specification we have the `deployment_path` too.

About support both environments. I think we can do a validation to see if both conda and docker are specified in MLProject and only proceed if just one of them are available. This way we avoid misleading behaviors related to it. Sounds good?

marcusrehm replied:
Hi @smurching , thank you for the feedback, I agree with you about put docker configuration inside MLProject, actually just the image, right now I don't see any reasons for explicit declare the mounting path. I used it but right now I don't see in training phase why we would need that.

Let's keep just the image attribute `dockerimage` inside MLProject file bellow `docker_env'. I'll use the suggestion you gave for the default mounting path `/mlflow/projects/code`.

About the run configuration I got it. We'll use the same approach as for conda environments. Good to keep the pattern.

The only question I got now is that for this first version, should we mount the local path or copy the project inside the container?

smurching replied:
> The only question I got now is that for this first version, should we mount the local path or copy the project inside the container?

@marcusrehm good question, I think we should copy the project inside the container. We can later add a `--no-build` flag to allow for mounting the local path instead, as described in [the proposal doc](https://docs.google.com/document/d/1GCFqMKUd4Ji4UMgYGeKW_afGVV-lUHwXF_RaQ7E_CpA/edit#)

marcusrehm commented:
Hi @smurching ! Thanks for replying! :)

I'll read the doc you shared and I'll try to contribute too.

About the comments above, I'll make the changes related do .gitignore and the validation today. Regarding the `run` tag and the docker.yaml usage I would like to know what you think about the points I described above.

marcusrehm commented:
Hi @smurching , did you have time to review these questions?

Thanks!

smurching reviewed (commented):
@marcusrehm yes! Here's a doc with an updated proposal after chatting with some committers: https://docs.google.com/document/d/1GCFqMKUd4Ji4UMgYGeKW_afGVV-lUHwXF_RaQ7E_CpA/edit#. Happy to discuss further & thanks for your patience.

Also left a couple of comments - if the proposal in the doc seems good to you, we should find a way to add tests to this PR. Specifically, we should add tests to `tests/projects/test_projects.py` verifying that `mlflow.projects.run()` runs successfully on a project with a docker env (to achieve this, we could add a copy of examples/docker to `tests/resources`) & that we raise appropriate exceptions on error cases (running a project with more than one env, running without `docker` on the PATH, running with `mode != "local").

marcusrehm commented on mlflow/projects/__init__.py:
```diff
@@ -72,15 +73,30 @@ def _run(uri, entry_point="main", version=None, parameters=None, experiment_id=N
+            _validate_docker_config(project.docker_config)
+            deployment_path = project.docker_config.get('run').get('deployment_path')
+            dockerimage = project.docker_config.get('run').get('dockerimage')
+            tracking.MlflowClient().set_tag(active_run.info.run_uuid, MLFLOW_DOCKER, True)
```
This one here is outdated but github is showing the same as before.

smurching reviewed (commented):
Thanks for the updates, leaving some comments. Also, would you be able to add some unit tests as described in https://github.com/mlflow/mlflow/pull/555#pullrequestreview-167690176? I think if we can address the comments & add tests this PR will be in pretty good shape :)

marcusrehm commented on mlflow/projects/__init__.py:
```diff
@@ -515,22 +512,91 @@ def _get_conda_command(conda_env_name):
+
+    dockerfile_template = (
+        "FROM {imagename}\n"
+        "LABEL Name={built_image} Version=0.0.1 AUTHOR=MarcusRehm\n"
```
Done.

marcusrehm commented on mlflow/projects/__init__.py:
```diff
@@ -515,22 +512,91 @@ def _get_conda_command(conda_env_name):
+                                        experiment_id=active_run.info.experiment_id).items():
+        dockerfile_template_env_vars += ["ENV {key}={value}\n".format(key=key, value=value)]
+
+    with TempDir() as tmp:
```
I switched the building process to use docker SDK, but for run I'd prefer to do it later because of extra logic in the `_run` related to mlflow execution just after the environment choice (docker or conda):
```python
        if block:
            command += _get_entry_point_command(project, entry_point, parameters, storage_dir)
            command = command_separator.join(command)
            return _run_entry_point(command, work_dir, exp_id, run_id=active_run.info.run_uuid)
        # Otherwise, invoke `mlflow run` in a subprocess
        return _invoke_mlflow_run_subprocess(
            work_dir=work_dir, entry_point=entry_point, parameters=parameters, experiment_id=exp_id,
            use_conda=use_conda, storage_dir=storage_dir, run_id=active_run.info.run_uuid)
```

marcusrehm commented on mlflow/projects/__init__.py:
```diff
@@ -515,22 +512,91 @@ def _get_conda_command(conda_env_name):
+            f.writelines(dockerfile_template)
+            f.writelines(dockerfile_template_env_vars)
+        eprint("building docker image")
+        eprint(dockerfile)
```
Done.

marcusrehm commented on mlflow/projects/__init__.py:
```diff
@@ -515,22 +512,91 @@ def _get_conda_command(conda_env_name):
+        with open(dockerfile, "w") as f:
+            f.writelines(dockerfile_template)
+            f.writelines(dockerfile_template_env_vars)
+        eprint("building docker image")
```
Done, but I kept the eprint for the message "=== Building docker image ===". Is it ok or should we use logging api to do that?

marcusrehm replied:
Done.

marcusrehm commented on mlflow/projects/__init__.py:
```diff
@@ -515,22 +512,91 @@ def _get_conda_command(conda_env_name):
+        eprint("building docker image")
+        eprint(dockerfile)
+        proc = Popen([docker_path, "build", "-t", built_image, "-f", "Dockerfile", "."],
+                    cwd=cwd,
```
Goit it! Done.

marcusrehm commented on mlflow/projects/__init__.py:
```diff
@@ -515,22 +512,91 @@ def _get_conda_command(conda_env_name):
+    dockerfile_template = (
+        "FROM {imagename}\n"
+        "LABEL Name={built_image} Version=0.0.1 AUTHOR=MarcusRehm\n"
+        "COPY {work_dir} {project_mount_path}\n"
```
Done.

marcusrehm replied:
Done, but Github didn't update it as Outdated.

marcusrehm commented on mlflow/projects/__init__.py:
```diff
@@ -515,22 +512,91 @@ def _get_conda_command(conda_env_name):
+    dockerfile_template_env_vars = []
+    for key, value in _get_run_env_vars(run_id=active_run.info.run_uuid,
+                                        experiment_id=active_run.info.experiment_id).items():
+        dockerfile_template_env_vars += ["ENV {key}={value}\n".format(key=key, value=value)]
```
Good, I agree with you @smurching , right now we can validate to run only if the tracking is a remote URI. For docker I think this would be the most common approach.

marcusrehm replied:
I think the first option would be better for now, don't you think? I'm validating if the tracking URI is local and if so it throws an exception say that it needs to be a remote URI.

smurching reviewed (commented):
Thanks for the updates, leaving some more comments! Let me know if there's anything I can do to help with adding tests / if you have any questions about that.

Edit: I think one tricky thing with the tests will be actually invoking `mlflow run <some-docker-project>`, since we require running against a remote URI for now. For this we can either 1) launch a local tracking server as we do in [existing REST tests](https://github.com/mlflow/mlflow/blob/master/tests/tracking/test_rest_tracking.py#L69) and use the `host.docker.internal` DNS endpoint to access the tracking server from within the container ([see docs](https://docs.docker.com/docker-for-windows/networking/#there-is-no-docker0-bridge-on-windows), this should work if Travis uses a recent-enough version of docker), or (2) add logic for mounting local `mlruns` directories inside the container & rely on that

marcusrehm commented on dev-requirements.txt:
```diff
@@ -8,3 +8,4 @@ coverage
 pypi-publisher
 scikit-learn
 scipy
+docker==3.6.0
```
Yes, you're right. removed.

marcusrehm replied:
Removed.

marcusrehm commented on examples/docker/conda_env.yaml:
```diff
@@ -0,0 +1,11 @@
+name: tutorial
```
OK, done.

marcusrehm commented on examples/docker/docker/Dockerfile:
```diff
@@ -0,0 +1,15 @@
+FROM continuumio/miniconda:4.5.4
```
The problem here is that the Dockerfile generated by mlflow is created inside the project folder, so if this Dockerfile is in `examples/docker/` it'll be overwritten. This Dockerfile was just to keep track of the  base image.

marcusrehm replied:
We can move it to examples\docker if we rename it something like Dockerfile.baseimage. What do you think?

smurching commented:
Thanks for the updates @marcusrehm, I think this PR is getting there but it'd be good to add some tests - if you'd like, I'd be happy to help write some & push them to this PR branch, or provide guidance on tests we should add.

marcusrehm commented:
Hi @smurching ! Thank you for your feedback!

Actually if you could help me with the tests would be nice. Unfortunately, as you could see, I had little time to work on this PR in the last weeks.

marcusrehm commented on mlflow/projects/__init__.py:
```diff
@@ -478,6 +498,94 @@ def _invoke_mlflow_run_subprocess(
+    if not docker_env.get('dockerimage'):
+        raise ExecutionException("No docker dockerimage specified.")
+    if mlflow.tracking.utils._is_local_uri(tracking.get_tracking_uri()):
+        raise ExecutionException("Right now docker environments support only remote tracking."
```
Done.

marcusrehm commented on mlflow/projects/__init__.py:
```diff
@@ -478,6 +498,94 @@ def _invoke_mlflow_run_subprocess(
+
+def _validate_docker_env(docker_env):
+    if not docker_env.get('dockerimage'):
+        raise ExecutionException("No docker dockerimage specified.")
```
Done.

marcusrehm commented on mlflow/projects/_project_spec.py:
```diff
@@ -22,6 +22,13 @@ def load_project(directory):
+    if not project_name:
+        project_name = None
+    docker_env = yaml_obj.get("docker_env")
+    if docker_env and not docker_env.get("dockerimage"):
```
Done.

marcusrehm commented on mlflow/projects/__init__.py:
```diff
@@ -478,6 +498,94 @@ def _invoke_mlflow_run_subprocess(
+    ``mlflow.projects.MLFLOW_DOCKER_HOME`` is unspecified, this method simply returns the passed-in
+    executable name.
+    """
+    docker_home = os.environ.get(MLFLOW_DOCKER_HOME)
```
Yes, I've done some researching and didn't find anything related to it. I removed that.

marcusrehm commented on mlflow/projects/__init__.py:
```diff
@@ -515,22 +512,91 @@ def _get_conda_command(conda_env_name):
+def _build_docker_image(work_dir, project, active_run, project_mount_path="/mlflow/projects/code"):
+    os.chdir(work_dir)
+    docker_path = _get_docker_bin_executable("docker")
+    built_image = "mlflow-{name}".format(
```
I created two tags for logging so we can have the image name and the image ID. I've putted the git version in image name too.

marcusrehm commented on mlflow/projects/__init__.py:
```diff
@@ -478,6 +498,94 @@ def _invoke_mlflow_run_subprocess(
+
+def _build_docker_image(work_dir, project, project_mount_path="/mlflow/projects/code"):
+    os.chdir(work_dir)
+    built_image = "mlflow-{name}".format(name=(project.name if project.name else "built-image"))
```
I've put some validations as you told, so if project has no name it'll raise an exception. Besides that I added the git version to the image name.

marcusrehm replied:
Now when mlflow builds the image it's named as "mlflow-{project-name-in-MLProject}-{version}". Where version is the git repo version. This way if the user doesn't change the commit, the image won't overwritten every in run.

Example: `mlflow-docker-example-13c3c47`

marcusrehm commented on examples/docker/MLproject:
```diff
@@ -0,0 +1,13 @@
+# conda_env: conda_env.yaml
+
+docker_env:
+  dockerimage:  mlflow-run-image-file
```
Hi @smurching , I agree with you. For this PR I think that in the context of mlflow the base image (which is generated by this Dockerfile) should be created before and not during run. We can support it later.

About Dockerfile in the root folder `examples/docker/Dockerfile` my concern here is that `mlflow run` also created its Dockerfile inside `examples/docker/` and after that the original file would be overwritten. To workaround it we can rename the base image Dockerfile to Dockerfile.baseimage.

What do you think about it?

marcusrehm commented on docs/source/projects.rst:
```diff
@@ -215,6 +219,17 @@ for your run. Then, run your project via
+.. code::
+
+    docker_env:
+        dockerimage: mlflow-run-image
```
The question here is that the image needs to have mlflow already installed. This is the reason I left the base image Dockerfile inside the docker folder.

smurching committed:
WIP

smurching committed:
WIP

smurching committed:
WIP

smurching committed:
Add support for running against Databricks & with local tracking URIs

smurching committed:
Update example README

smurching committed:
WIP

smurching committed:
Fix tests & lint

smurching commented:
@marcusrehm thanks for the updates! I just pushed some unit tests & doc updates, plus some logic for propagating local & Databricks tracking URIs inside the docker container. I'd welcome any feedback on my updates, will also get someone else to make a review pass - AFAICT the functionality otherwise looks pretty good :)

marcusrehm commented:
Hi @smurching ! Thank you very much for the help!

I'm doing the review and testing just to double check, but it seems pretty cool!

smurching committed:
Actually add docker tests & clean up docs a bit

smurching committed:
Code cleanups

smurching commented:
Oops, realized my last commit didn't actually contain the tests - just pushed them :P

smurching committed:
Rename test resource file to avoid name conflicts

smurching committed:
Fix test in travis

Commit:
remove unneeded project validation and improve docker example readme

smurching committed:
Raise appropriate exception when attempting to run docker project on Databricks

Commit:
enables building docker images in Windows OS

Commit:
Fix lint

Commit:
Remove specific exception handle only for python 3 when deleting temp docker files

Commit:
improve handling exception when deleting docker temp files

Commit:
lint

Commit:
lint  disable=broad-except in removing temp docker files

Commit:
lint tests

Commit:
build docker example base image before tests

marcusrehm commented:
Hi @smurching , I've did some updates and reviewed the tests as you asked. The feature seams pretty good now. Do we need anything more in order to close? What do you think?

smurching commented:
@marcusrehm it looks good to me! Thanks so much for all the hard work over the past weeks, let's merge this - looking forward to seeing it go out with 0.9.0 :)!!

smurching commented:
Shoutout also to @mparkhe for making a pass over the PR :)

marcusrehm commented:
Cool!!! Thanks guys for all the help!
